### ISSUE TITLE
After train nuscence  6000 iterations with batchsize of 16  , loss and grad_norm become nan.  use conf  stream_petr_r50_flash_704_bs2_seq_90e.py

### ISSUE THREAD

**Issue by lianshushu** [exiawsh/StreamPETR#17]
2023-05-22 21:39:18,645 - mmdet - INFO - Iter [7050/79110]      lr: 7.844e-04, eta: 2 days, 10:28:21, time: 1.844, data_time: 0.164, memory: 21441, frame_0_loss_cls: 1.2109, frame_0_loss_bbox: 2.0770, frame_0_d0.loss_cls: 1.2623, frame_0_d0.loss_bbox: 2.0982, frame_0_d1.loss_cls: 1.2512, frame_0_d1.loss_bbox: 2.0969, frame_0_d2.loss_cls: 1.2396, frame_0_d2.loss_bbox: 2.0894, frame_0_d3.loss_cls: 1.2348, frame_0_d3.loss_bbox: 2.0902, frame_0_d4.loss_cls: 1.2180, frame_0_d4.loss_bbox: 2.0775, frame_0_dn_loss_cls: 0.9295, frame_0_dn_loss_bbox: 1.7723, frame_0_d0.dn_loss_cls: 0.9603, frame_0_d0.dn_loss_bbox: 1.7837, frame_0_d1.dn_loss_cls: 0.9425, frame_0_d1.dn_loss_bbox: 1.7803, frame_0_d2.dn_loss_cls: 0.9305, frame_0_d2.dn_loss_bbox: 1.7719, frame_0_d3.dn_loss_cls: 0.9298, frame_0_d3.dn_loss_bbox: 1.7708, frame_0_d4.dn_loss_cls: 0.9286, frame_0_d4.dn_loss_bbox: 1.7733, frame_0_enc_loss_cls: nan, frame_0_enc_loss_bbox: nan, frame_0_enc_loss_iou: nan, frame_0_centers2d_losses: nan, frame_0_centerness_losses: nan, loss: nan, grad_norm: nan

**Comment by exiawsh** [comment]
Hi, I think there is something wrong, becasue the total iteration is 79110 in your long. In my training log, the iteration is 158220. I need more details about the training config and the number of gpus.

**Comment by lianshushu** [comment]
yes,I use 2 GPU.
main conf info is :
num_gpus = 2
batch_size = 16
num_iters_per_epoch = 28130 // (num_gpus * batch_size)
num_epochs = 90
lr_config = dict(
    policy='CosineAnnealing',
    warmup='linear',
    warmup_iters=500,
    warmup_ratio=1.0 / 3,
    min_lr_ratio=1e-3,
    )

**Comment by exiawsh** [comment]
> yes,I use 2 GPU. main conf info is : num_gpus = 2 batch_size = 16 num_iters_per_epoch = 28130 // (num_gpus * batch_size) num_epochs = 90 lr_config = dict( policy='CosineAnnealing', warmup='linear', warmup_iters=500, warmup_ratio=1.0 / 3, min_lr_ratio=1e-3, )

Try these two configs:
1.
num_gpus = 2
batch_size = 8
num_iters_per_epoch = 28130 // (num_gpus * batch_size)
num_epochs = 90
lr_config = dict(
policy='CosineAnnealing',
warmup='linear',
warmup_iters=500,
warmup_ratio=1.0 / 3,
min_lr_ratio=1e-3,
)

2.
num_gpus = 2
batch_size = 16
num_iters_per_epoch = 28130 // (num_gpus * batch_size)
num_epochs = 90

optimizer = dict(
    type='AdamW', 
    **lr=6e-4, # bs 8: 2e-4 || bs 16: 4e-4**
    paramwise_cfg=dict(
        custom_keys={
            'img_backbone': dict(lr_mult=0.25), # 0.25 only for Focal-PETR with R50-in1k pretrained weights
        }),
    weight_decay=0.01)

lr_config = dict(
policy='CosineAnnealing',
warmup='linear',
warmup_iters=500,
warmup_ratio=1.0 / 3,
min_lr_ratio=1e-3,
)

**Comment by happygds** [comment]
Hi @exiawsh , how do you suggest the setting of learning rate when I train model with the batchsize of 16/32 on A100? Otherwise, the small batch would waste a lot of memory.

**Comment by exiawsh** [comment]
> Hi @exiawsh , how do you suggest the setting of learning rate when I train model with the batchsize of 16/32 on A100? Otherwise, the small batch would waste a lot of memory.

total batch_size 16: lr 4e-4
total batch_size 32: lr6e-4
I have not tried setting  with the total batch size 64 yet, and I believe that setting lr between 6e-4 and 1e-3 may converge.

**Comment by happygds** [comment]
> > Hi @exiawsh , how do you suggest the setting of learning rate when I train model with the batchsize of 16/32 on A100? Otherwise, the small batch would waste a lot of memory.
> 
> total batch_size 16: lr 4e-4 total batch_size 32: lr6e-4 I have not tried setting with the total batch size 64 yet, and I believe that setting lr between 6e-4 and 1e-3 may converge.

If I use total bachsize of 128, maybe I should try something like LAMB optimizer ?

**Comment by exiawsh** [comment]
@happygds Sorry, I haven't tried before. If you feel that memory usage is a bit wasteful and want to speed up training, you can set with_cp to False in the config. (which can speed up about 10-20%, the memory usage will be at least 2-3x)

**Comment by happygds** [comment]
> @happygds Sorry, I haven't tried before. If you feel that memory usage is a bit wasteful and want to speed up training, you can set with_cp to False in the config. (which can speed up about 10-20%, the memory usage will be at least 2-3x)

Thanks a lot !
